src/propagators.jl

```diff
@@ -0,0 +1,325 @@
+using QuBase
+using Docile
+
+import Base: start, next, done
+
+abstract QuantumPropagator{B<:AbstractBasis,T,A}
+
+type StepPropagator{B<:AbstractBasis,T,A} <: QuantumPropagator{B,T,A}
+    hamiltonian::QuMatrix{B,T,A}
+    init_state::QuVector
+    tlist::Array
+    StepPropagator(hamiltonian,init_state,tlist) = new(hamiltonian,init_state,tlist)
+end
+
+
+@doc """
+Euler Problem
+`hamiltonian` : Hamiltonian of the system
+`init_state`  : initial state of the syste
+`tlist`       : Time step array
+""" ->
+immutable QuEuler{B<:AbstractBasis,T,A} <: QuantumPropagator{B,T,A}
+    hamiltonian::QuMatrix{B,T,A}
+    init_state::QuVector
+    tlist::Array
+    QuEuler(hamiltonian, init_state, tlist) = new(hamiltonian, init_state, tlist)
+end
+
+
+QuEuler{B<:AbstractBasis,T,A}(hamiltonian::QuMatrix{B,T,A}, init_state::QuVector, tlist::Array)=QuEuler{B,T,A}(hamiltonian, init_state, tlist)
+
+
+@doc """
+Euler State
+`psi`     : State of the system
+`t`       : Current time in time step `tlist` array
+`t_state` : Index of the next time state in the `tlist` array
+""" ->
+immutable QuEulerState
+    psi::QuVector
+    t
+    t_state
+end
+
+
+@doc """
+Input Parameters : Euler Problem
+
+Returns the starting, i.e., the initial state of the system
+""" ->
+function Base.start(prob::QuEuler)
+    t_state = start(prob.tlist)
+    t,t_state = next(prob.tlist,t_state)
+    return QuEulerState(prob.init_state,t,t_state)
+end
+
+
+@doc """
+Input Parameters : Euler Problem and Euler State
+
+Returns the next state using Euler's method.
+""" ->
+function Base.next(prob::QuEuler, qustate::QuEulerState)
+    op = prob.hamiltonian
+    t_list = prob.tlist
+    current_qustate = qustate.psi
+    current_t = qustate.t
+    if current_t != t_list[end]
```

Comment:
I think we don't need this check. The iterator protocol should make sure that `next` is not called if `done` returns false. And even if it would be called it would just return the last element of the list and `dt==0`. But we might have to make sure that this is actually true.